pytorch_toolkit/human_pose_estimation/datasets/transformations.py

```diff
@@ -413,3 +414,141 @@ def __call__(self, sample):
+
+        points = np.array([center, center + shift_x, center + shift_y], dtype=np.float32)
+        transformed_points = np.array([[output_size[0] * 0.5, output_size[1] * 0.5],
+                                       [0, output_size[1] * 0.5],
+                                       [output_size[0] * 0.5, output_size[1] * 0.5 - output_size[0] * 0.5]],
+                                       dtype=np.float32)
+
+        rev_trans = cv2.getAffineTransform(np.float32(transformed_points), np.float32(points))
+
+        trans = cv2.getAffineTransform(np.float32(points), np.float32(transformed_points))
+
+        return trans, rev_trans
+
+    @staticmethod
+    def _affine_transform(pt, t):
+        new_pt = np.array([pt[0], pt[1], 1.])
+        new_pt = np.dot(t, new_pt)
+        return new_pt[:2]
+
+    @staticmethod
+    def _rotation(point, r):
+        r = np.pi * r / 180
+        return [point[0] * np.cos(r) - point[1] * np.sin(r), point[0] * np.sin(r) + point[1] * np.cos(r)]
+
+
+class HalfBodyTransform(object):
+    def __init__(self, aspect_ratio=0.75, prob=0.3):
+        self._prob = prob
+        self.aspect_ratio = aspect_ratio
+        self._num_keypoints = None
+    def __call__(self, sample):
+        rand = np.random.uniform()
+        do_body_transform = rand <= self._prob and np.sum(sample['keypoints'][2:][::3]) > self._num_keypoints
+        if not do_body_transform:
+            return sample
+
+        upper_points = []
+        lower_points = []
+        for idx in range(len(sample['keypoints']) // 3):
+            if sample['keypoints'][idx * 3 + 2] > 0:
+                if idx < 11:
+                    upper_points.append([sample['keypoints'][idx * 3], sample['keypoints'][idx * 3 + 1]])
+                else:
+                    lower_points.append([sample['keypoints'][idx * 3], sample['keypoints'][idx * 3 + 1]])
+
+        if np.random.uniform() < 0.5 and len(upper_points) > 2 or len(lower_points) < 2:
+            target_points = upper_points
+        else:
+            target_points = lower_points
+
+        if len(target_points) < 2:
+            center, scale = None, None
+        else:
+            target_points = np.array(target_points, dtype=np.float32)
+            center = np.array([target_points[:, 0].sum() / len(target_points),
+                               target_points[:, 1].sum() / len(target_points)], dtype=np.float32)
+
+            w = target_points[:, 0].max() - target_points[:, 0].min()
+            h = target_points[:, 1].max() - target_points[:, 1].min()
+
+            if w > self.aspect_ratio * h:
+                h = w / self.aspect_ratio
+            elif w < self.aspect_ratio * h:
+                w = h * self.aspect_ratio
+
+            scale = np.array([w / 200, h / 200], dtype=np.float32)
+            scale *= 1.5
+
+        if center is not None:
+            sample['center'] = center
+            sample['scale'] = scale
+
+        return sample
+
+
+class Normalization(object):
```

Comment:
No, it's not very convenient here because all transforms class get dict `sample`,  not image. In addition, normalization should be before ChannelPermutation.